services/director-v2/src/simcore_service_director_v2/api/routes/dynamic_services.py

```diff
@@ -332,3 +337,44 @@ async def update_projects_networks(
         rabbitmq_client=rabbitmq_client,
         project_id=project_id,
     )
+
+
+@router.get(
+    "/projects/{project_id}/inactivity", summary="returns if the project is inactive"
+)
+@log_decorator(logger=logger)
+async def get_project_inactivity(
+    project_id: ProjectID,
+    max_inactivity_seconds: NonNegativeFloat,
+    scheduler: Annotated[DynamicSidecarsScheduler, Depends(get_scheduler)],
+    projects_repository: Annotated[
+        ProjectsRepository, Depends(get_repository(ProjectsRepository))
+    ],
+) -> GetProjectInactivityResponse:
+    project: ProjectAtDB = await projects_repository.get_project(project_id)
+
+    inactivity_responses: list[InactivityResponse] = await logged_gather(
+        *[
+            scheduler.get_service_inactivity(NodeID(node_id))
+            for node_id in project.workbench
+            # NOTE: only new style services expose service inactivity information
+            # director-v2 only tracks internally new style services
+            if scheduler.is_service_tracked(NodeID(node_id))
+        ]
+    )
+
+    # NOTE: if a service results in being active it means one of the following
+    # - it does not provide information about it's inactivity state
+    # - the service is actually used
+    inactive_services: list[InactivityResponse] = [
+        r for r in inactivity_responses if r.is_inactive
+    ]
+
+    # a project is considered inactive when all it's inactive services are inactive for
+    # more than `max_inactivity_seconds`
```

Comment:
so that means if 1 service is useless for a longer time, it will not be shutdown? only when all of them running services in a project? why is that?

Reply:
I'll fix the wording. Sorry for that, just read it again.

The API to figure out service inactivity is only called when the frontend detects that the user has been inactive for 30 minutes.
If one service is still active, the frontend will wait. When all services are inactive, the fronted knows it is safe to close this.

Reply:
I get that. but why not switching off single services? that is the part I do not understand.

Reply:
Because it's not this endpoint's responsibility to turn off/shutdown the services.
The shut-down is done by the frontend.